sabnzbd/nzbparser.py

```diff
@@ -54,98 +49,123 @@ def nzbfile_parser(raw_data: str, nzo):
+
+                    # Sort the articles by part number, compatible with Python 3.5
+                    raw_article_db_sorted = [raw_article_db[partnum] for partnum in sorted(raw_article_db)]
+
+                    # Create NZF
+                    nzf = sabnzbd.nzbstuff.NzbFile(file_date, file_name, raw_article_db_sorted, file_bytes, nzo)
+
+                    # Check if we already have this exact NZF (see custom eq-checks)
+                    if nzf in nzo.files:
+                        logging.info("File %s occured twice in NZB, skipping", nzf.filename)
+                        continue
+
+                    # Add valid NZF's
+                    if file_name and nzf.valid and nzf.nzf_id:
+                        logging.info("File %s added to queue", nzf.filename)
+                        nzo.files.append(nzf)
+                        nzo.files_table[nzf.nzf_id] = nzf
+                        nzo.bytes += nzf.bytes
+                        valid_files += 1
+                        avg_age_sum += file_timestamp
                     else:
-                        raw_article_db[partnum] = (article_id, segment_size)
-                        file_bytes += segment_size
-                except:
-                    # In case of missing attributes
-                    pass
-
-        # Sort the articles by part number, compatible with Python 3.5
-        raw_article_db_sorted = [raw_article_db[partnum] for partnum in sorted(raw_article_db)]
-
-        # Create NZF
-        nzf = sabnzbd.nzbstuff.NzbFile(file_date, file_name, raw_article_db_sorted, file_bytes, nzo)
-
-        # Check if we already have this exact NZF (see custom eq-checks)
-        if nzf in nzo.files:
-            logging.info("File %s occured twice in NZB, skipping", nzf.filename)
-            continue
-
-        # Add valid NZF's
-        if file_name and nzf.valid and nzf.nzf_id:
-            logging.info("File %s added to queue", nzf.filename)
-            nzo.files.append(nzf)
-            nzo.files_table[nzf.nzf_id] = nzf
-            nzo.bytes += nzf.bytes
-            valid_files += 1
-            avg_age_sum += file_timestamp
-        else:
-            logging.info("Error importing %s, skipping", file_name)
-            if nzf.nzf_id:
-                sabnzbd.remove_data(nzf.nzf_id, nzo.admin_path)
-            skipped_files += 1
+                        logging.info("Error importing %s, skipping", file_name)
+                        if nzf.nzf_id:
+                            sabnzbd.remove_data(nzf.nzf_id, nzo.admin_path)
+                        skipped_files += 1
+                    element.clear()
+        except:
+            # Remove all data added to the nzo
+            for nzf in nzo.files:
+                nzf.remove_admin()
+            nzo.first_articles = []
+            nzo.first_articles_count = 0
+            nzo.bytes_par2 = 0
+            nzo.files = []
+            nzo.files_table = {}
+            nzo.bytes = 0
```

Comment:
I saw you added this, but why?
It breaks the `allow_incomplete_nzb` functionality. I know that's a weird one, but some users wanted it.
Additionally, if `allow_incomplete_nzb` is disabled, the `nzo.purge_data()` in `NzoObject.__init__` will clear all the stuff out, right?

Reply:
I'm not entirely sure if it should be there. This part of the parsing doesn't work quite like before. In the old version, if the XML parser found a problem then it would throw an exception and the import loop would never start so the `nzo` would always be empty.

Now that I think of it, I think it shouldn't be there. I added it while working on the regex parser and then I figured I needed it here as well because both import the file incrementally. However, the reason why it was necessary in the regex version was that the XML parser would use the same `nzo`.